### PR TITLE
KEYCLOAK-17368 Show forwarded errors when a default remote IdP is configured

### PR DESCRIPTION
**Pull request by lscorcia**
Scenario: a remote SAML IdP is configured as a default Identity Provider in the Authentication - Identity Provider Redirector configuration. 

When a remote IdP returns a SAML error (AuthnFailed), the error status code and description are returned to the IdP and generally they are shown to the user as an explanation of what went wrong.

When a default IdP is set in the Identity Provider Redirector however, any remote failure messages are not displayed to the user and the authentication process loops, returning the user to the remote IdP over and over again with no explanation.

This PR stops the default redirect when an error is forwarded, and shows the error message to the user.

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java:68]: Will it be better to use "context.challenge" here? I am not 100% sure, but ideally this would mean that:
- In case that IdentityProviderAuthenticator is REQUIRED, the challenge will be forced to be displayed
- In case that IdentityProviderAuthenticator is ALTERNATIVE, the challenge won't be displayed, but it will go to the next authenticator (usually username&password screen) and the error will be displayed on this screen. Which will give user a chance to fallback to the authentication with username&password . WDYT?

**Review by mposolda — CHANGES_REQUESTED** [review]
@lscorcia Thanks a lot for the PR! I think it is good step moving forward.

I've added one discussion comment inline. Once we agree on the proposed path with that, we may need also the automated test for this bugfix.

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java
Comment (lscorcia) [reply services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java:68]: I agree your proposal is better, however after trying it seems that the `context.challenge` call the control is immediately passed to the challenge (error) page and it never gets a chance to reach the forms authenticator.

I have also tried replacing this code with a simple context.attempted(), when ALTERNATIVE this actually redirects to the forms authenticator and shows the error message, but when REQUIRED it goes directly to the "Invalid username or password" screen and does not show the forwarded error. Maybe the issue is that forwarded messages are lost when the flow ends without any successful execution?

Comment (mposolda) [reply services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java:68]: Hm... I am not 100% sure. It is possible that it is a bug as "context.challenge" should ideally work in same way as "context.attempted" . The only difference is, that "context.challenge" provides a way for the authenticator to provision the screen when the "context.attempted" is mostly targetted for non-interactive authenticators when showing any specific screen is usually not desired. But maybe I am missing something...

If you want to investigate the authenticationFlow (and possibly fix at the authenticationFlow level), it will be ideal. But if not, it is ok for now to keep things as you did in PR. Just maybe:
- add TODO in the code and possibly refer to this PR (or summary in the TODO if possible)
- Additional very minor thing: Could you please use:
```
if (something) {
```
instead of:
```
if (something)
{
```

And finally the test :-)
Thanks!

Comment (lscorcia) [reply services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java:68]: Yep, the tests are always the hardest part. I still have no idea how to setup the realms and config, usually I just go look for some test similar to the one I need and start to tweak, but it takes ages. Will try anyway! :)

Regarding the .challenge issue, I think it's because the flow executes the challenge with the same priority of the authenticator, and the IdP auth has a greater priority than the forms auth. I'll check if I can change the challenge priority. Any suggestions welcome, and thanks for reviewing this!

**Comment by lscorcia** [comment]
I spent some more time trying to learn the authentication flow mechanism. Since any challenge is returned before moving to the next available authenticator, it seems that the more appropriate approach is to let the Identity Provider declare it has been "attempted" when a forwarded error is received. This lets the next available authenticator display any forwarded error message.

If there is no other authenticator available or the current authenticator is required, the authentication processor currently throws an exception that presents a generic error message and does not display the details of any forwarded error. 

I have added a simple check before generating the exception: if there is any forwarded error message, it shows that instead of the generic error message.

Now the hard part: tests. I really have no idea where to start. I have found no tests for the Default Idp feature and I have no idea how to setup the required scenario. Any suggestion or reference to similar tests would be really welcome as I am pretty much stuck. Thanks!

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java
Comment (lscorcia) [reply services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java]: Updated with a slightly different approach. I'll mark the thread as resolved and wait for your kind review on the other one.

**Comment by mposolda** [comment]
@lscorcia Your approach makes sense to me. Thanks for your investigation and the changes!

Regarding the test, I agree that it sometimes can be a bit harder. Especially with regards to identityBrokering and authenticator SPI. There is KcOidcBrokerIdpHintTest, which tests some aspects of IdentityProviderAuthenticator, but just in the context of "kc_idp_hint" parameter. It seems we don't have test for default idp set at the authenticator configuration level :( I suppose it may be good to introduce new test class in the "broker" package, which will inherit from AbstractInitializedBaseBrokerTest (something like KcOidcDefaultIdpTest), and which will focus on the configuration of default idp in the IdentityProviderAuthenticator.

For the tests, which tests the authenticator changes, there is for example BrowserFlowTest. See the usage of "FlowUtil" class from this test for the inspiration. Another option is to look at AuthenticatorConfigTest, which tests CRUD of the authenticator configurations for some existing authenticator execution - which is maybe what you need.

Hope this helps and good luck!

**Comment by mposolda** [comment]
@lscorcia Any update?

**Comment by lscorcia** [comment]
Hi @mposolda , thanks for checking in. I'm really sorry, but I had to pospone work on these tests due to other priorities. I still need this fixed, but I won't be able to work on it for the next weeks.

**Comment by lscorcia** [comment]
Hello @mposolda , I have prepared some basic tests for the Default IdP feature in https://github.com/keycloak/keycloak/pull/8157 using the tips you provided above. I would appreciate if you could take a look at it. 
If my approach is correct hopefully it shouldn't be too hard to also provide the tests for this issue!

**Comment by lscorcia** [comment]
Over the weekend my virtualbox instance crashed hard and it corrupted the virtual hard disk. It took a few unpleasant days to recover data from the virtual hard disk, so to avoid losing this work I now committed the whole thing. I will rebase it and remove the conflicts after #8157 gets merged.

**Comment by mposolda** [comment]
@lscorcia Sorry to hear the troubles you had with the virtual hard disk. I can imagine how unpleasant this is... I've merged your other PR #8157. Thanks a lot for adding the tests in that PR! Do you please have a chance to rebase this and fix the conflicts?

**Comment by lscorcia** [comment]
Thank you @mposolda ! Rebased and pushed.

**Comment by lscorcia** [comment]
Hi @mposolda , do you have time to review this?

**Comment by mposolda** [comment]
@lscorcia Sorry for late review. The PR looks good to me. Do you please have a chance to rebase? Hopefully it will help also with the errors in `JavascriptAdapterTest` (I think those are not related to this PR, but I am not 100% sure and I see they failed in both quarkus and wildfly).

**Comment by lscorcia** [comment]
Hello @mposolda , rebased as requested! Thanks for your time!

**Review by mposolda — APPROVED** [review]
@lscorcia Thanks!

**Comment by peterzandbergen** [comment]
Hi, has this change made it into a release yet? Cheers

**Pull request merged**
